clash-lib/src/Clash/Core/Util.hs

```diff
@@ -932,3 +933,73 @@ tySym
 tySym m (coreView1 m -> Just ty) = tySym m ty
 tySym _ (LitTy (SymTy s))        = return s
 tySym _ ty = throwE $ $(curLoc) ++ "Cannot reduce to a string:\n" ++ showPpr ty
+
+-- | Determine whether we should split away types from a product type, i.e.
+-- clocks should always be separate arguments, and not part of a product.
+shouldSplit
+  :: TyConMap
+  -> Type
+  -- ^ Type to examine
+  -> Maybe (Term,[Type])
+  -- ^ If we want to split values of the given type then we have /Just/:
+  --
+  -- 1. The (type-applied) data-constructor which, when applied to values of
+  --    the types in 2., creates a value of the examined type
+  --
+  -- 2. The arguments types of the product we are trying to split.
+  --
+  -- Note that we only split one level at a time (although we check all the way
+  -- down), e.g. given /(Int, (Clock, Bool))/ we return:
+  --
+  -- > Just ((,) @Int @(Clock, Bool), [Int, (Clock, Bool)])
+  --
+  -- An outer loop is required to subsequently split the /(Clock, Bool)/ tuple.
+shouldSplit tcm = shouldSplit0 tcm . tyView . coreView tcm
+
+-- | Worker of 'shouldSplit', works on 'TypeView' instead of 'Type'
+shouldSplit0
+  :: TyConMap
+  -> TypeView
+  -> Maybe (Term,[Type])
+shouldSplit0 tcm (TyConApp tcNm tyArgs)
+  | Just tc <- lookupUniqMap tcNm tcm
+  , [dc] <- tyConDataCons tc
+  , let dcArgs  = substArgTys dc tyArgs
+  , let dcArgVs = map (tyView . coreView tcm) dcArgs
+  = if any (\ty -> isJust (shouldSplit0 tcm ty) || splitTy ty) dcArgVs then
+      Just (mkApps (Data dc) (map Right tyArgs), dcArgs)
+    else
+      Nothing
+ where
+  -- Currently we're only interested in splitting of Clock, Reset, and Enable
+  splitTy (TyConApp tcNm0 _)
+    = nameOcc tcNm0 `elem` [ "Clash.Signal.Internal.Clock"
+                           , "Clash.Signal.Internal.Reset"
+                           , "Clash.Signal.Internal.Enable"
```

Comment:
Are enable lines special in Verilog/VHDL?

Reply:
No, just like clock and reset aren't. However, ASIC synthesis tools derive clock gating from enables, so let's make sure they can do their job as easy as possible.

Reply:
Right, makes sense.